### PR TITLE
Improve canvas lti integration

### PR DESCRIPTION
**Pull request by andrew-gardener**
- Switch user id mapping from one field to a list (will using first existing non-empty field from parameter list for webwork user id)
- Added course title parser to config file for more flexibility in parsing course title
  - Set default to using custom_term_name."_".context_label (canvas term+course code)
  - will no longer use context_title since it can easily be very long (40+ characters) which can confuse users due to title cutoff
- removed storing user identifier for automatic syncing tasks (instead just pass admin)
- prevent system error when custom_homework_set or custom_quiz_set point to non-existent homework sets (user will still see relevant message in UI)
- added global lti_contexts table for storing links between consumer keys+context ids and webwork courses (along with membership and grade pass back info)
- removed `launch_auto_manage_user_roles` (instead just auto update all users. to much complexity for something to block users when assignment due date should be enough)
- automatically set webwork setting courseTitle to the context title

### CODE REVIEW

File: conf/localOverrides.conf.ubc.dist [hunk 1]
```diff
@@ -493,20 +497,17 @@ $bridge{vista_blacklist} = (
 # - Allowed characters in regex form: [a-zA-Z0-9_-]
 # - Max length 40 characters
 # See WebworkBridge's Parser::sanitizeCourseName() for implementation
-#
-# Maping Rules:
-# 1. Rule can be either extact match or regex
-# 2. Rules has priority, first come first match. If webwork got a match,
-#    it will use the match and the rules after that will not get evaluated.
-# 3. Using regex grouping for substituion
-$bridge{course_mapping_rules} = [
```
Comment (xcompass) [inline conf/localOverrides.conf.ubc.dist]: we still need this for blackboard as it is not retired for next term.

File: lib/WebworkBridge/Parser.pm [hunk 2]
```diff
@@ -43,43 +43,22 @@ sub parseCourse
...
-			if ($course =~ $regex)
-			{
-				# get the actual value of mapping if we don't need to eval it
-				my $cname = ($need_eval) ? eval($href->{$key}) : $href->{$key};
-				my $ret = sanitizeCourseName($cname);
-				debug("Using mapping '$key' for course '$course' to '$ret'");
-				return $ret;
-			}
-		}
+	my $course_title;
```
Comment (xcompass) [inline lib/WebworkBridge/Parser.pm]: same here, we need a backward compatible update

File: conf/localOverrides.conf.ubc.dist (hunk 1 above)
Comment (andrew-gardener) [reply conf/localOverrides.conf.ubc.dist]: Added it back as another function in `localOverrides.conf.ubc`

https://github.com/ubc/webwork2/pull/5/files#diff-852a279db2450e7d5ec105a9092706e5R521

This should allow use to easily remove the dependency by just editing the conf when blackboard is retired.

File: lib/WebworkBridge/Parser.pm (hunk 2 above)
Comment (andrew-gardener) [reply lib/WebworkBridge/Parser.pm]: see above

**Comment by andrew-gardener** [comment]
Also moved lti secrets to a hash in `$bridge{lti_secrets}` so that others can't simply link link to webwork using a key/secret that is another known setting in bridge (ex: key:`course_template` secret:`modelCourse`)

File: conf/localOverrides.conf.ubc.dist (hunk 1 above)
Comment (xcompass) [reply conf/localOverrides.conf.ubc.dist]: 👍

**Pull request merged**